src/scanner.ts

```diff
@@ -646,45 +646,53 @@ export class Scanner {
         let start = this.pos;
         while (true) {
             if (this.pos >= this.len) {
-                result += this.text.substring(start, this.pos);
+                result += this.text.slice(start, this.pos);
                 this.setTokenAsUnterminated();
                 this.getDiagnostics().report(this.pos, diagnostic || Diagnostics.Unterminated_string_literal);
                 break;
             }
 
             let ch = this.text.charCodeAt(this.pos);
+            let chFromEntity = false;
             if (decodeEscapeSequences && ch === CharacterCodes.Ampersand) {
-                result += this.text.substring(start, this.pos);
+                result += this.text.slice(start, this.pos);
                 this.pos++;
                 ch = this.scanCharacterEntity();
+
+                // Decoding escape sequences breaks assumptions about character start position.
+                // Set pos to the end of the escape sequence but slice start one character later,
+                // and record that we are in such a state.
                 start = this.pos;
+                this.pos--;

```

Comment:
I think the general problem is that we defer incrementing `this.pos` until later, while in `scan` we increment immediately after reading the character. This results in the "off-by-one" error we're seeing later. We might want to refactor the method to avoid backtracking `this.pos`, as I am concerned this could put the scanner in a bad state somewhere down the line.

Reply:
Can you add a test case to `src\tests\resources\specs\test.grammar` for this?

Reply:
Done.